### ISSUE TITLE
New lint: `doc_hidden_item_used`

### ISSUE THREAD

**Issue by sgued** [rust-lang/rust-clippy#11536]
### What it does

This lint triggers when a item declared with `#[doc(hidden)]` is used outside of the crate where the item is declared **and** outside of a macro expansion.

### Advantage

Most `#[doc(hidden)]` items are not meant to be part of the public API and are only there to allow macros to access "internals" of the crate.
As such, using such an item is generally a bad idea.

### Drawbacks

There can be cases where such an item is still part of the public API but hidden. The only case I know of that is serde's [`deserialize_in_place`](https://docs.rs/serde/1.0.188/src/serde/de/mod.rs.html#566).

I also know that [TypedBuilder](https://docs.rs/typed-builder) uses `#[doc(hidden)]` on the builder type, but I believe the lint should not trigger in that case. The builder is obtained with the documented `pub fn builder()` method, and the methods of the builder themselves are not `#[doc(hidden)]`. Since the builder type is never explicitly referenced when used, this lint should not trigger.

### Example

In crate `crate_a`:

```rust
#[doc(hidden)]
struct A;
```

In crate *B*: 

```rust
fn main() {
    // This triggers the lint.
    let a = crate_a::A;
}

**Comment by sgued** [comment]
For example, `cargo-semver-checks` does not consider `#[doc(hidden)]` items to be part of the public API: https://github.com/obi1kenobi/cargo-semver-checks/pull/576

**Comment by obi1kenobi** [comment]
I'm in favor of such a lint existing. However, we will have to be careful with edge cases, and about whether this lint is enabled by default or `pedantic` or something else.

One non-obvious edge case: `doc(hidden)`-ness is path-dependent, not just an inherent property of items. It's possible that one way of accessing an item is public API whereas another one is public but hidden: https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/pub.20re-export.20of.20item.20from.20.60doc.28hidden.29.60.20module

I plan on writing a post [on my blog](https://predr.ag/blog) about some of these edge cases in the near future.

**Comment by sgued** [comment]
Maybe it would make sense to split this lint in 2:

- `doc_hidden_item_used` in `restriction` therefore allow by default
- `non_public_item_used` in `suspicious` and therefore warn by default. This Would lint uses of items that are `#[doc(hidden)]` and that have another hint that they are private: starting with an underscore and/or having `private` in the name.

**Comment by obi1kenobi** [comment]
It might also be worth considering the broader picture of suspicious `doc(hidden)` use + suspicious uses of hidden items, and deciding how to go forward on clippy lints.

Here are a few examples of what I'd consider suspicious `doc(hidden)` use, taken from [the cargo-semver-checks test suite](https://github.com/obi1kenobi/cargo-semver-checks/pull/576).

#### Hidden variant in public API exhaustive enum

```rust
pub enum Example {
    First,

    #[doc(hidden)]
    Second,
}
```

I think this should lint by default and suggest making the enum `#[non_exhaustive]`. Otherwise, the public API is that "exhaustive matching is okay but `Example::Second` is not public API" which seems self-contradictory.

#### Hidden pub field in a public API exhaustive struct

```rust
pub struct Example {
    pub x: i64,

    #[doc(hidden)]
    pub y: i64,
}
```

Pub structs where all fields are pub are constructible using struct literals: `Example { x: 1, y: 2 }`, and can also be pattern-matched exhaustively as well. If the `y` field isn't meant to be public API, then this struct should _probably_ be non-exhaustive for the same reasons as above.

Equivalent cases to this one exist for:
- hidden pub fields inside public API tuple structs
- hidden pub fields of exhaustive public API struct variants inside enums
- hidden pub fields of exhaustive public API tuple variants inside enums

#### Public API field inside struct / enum variant is of `doc(hidden)` type

```rust
pub struct Example {
    pub x: Hidden,
}

#[doc(hidden)]
pub struct Hidden;
```

This is another suspicious and confusing API situation: users are allowed to access the field which under normal type privacy rules means the type must be public, and yet `doc(hidden)` says the type and its contents are not public. Which is it?

It raises some interesting edge cases:
- If the hidden type implements a trait that is public API, are the trait's associated functions considered public API here?
- If above you said "no, because `impl MyTrait for Type` is hidden if the type is hidden," then what if the trait had a blanket impl for all `T` — then arguably that blanket impl would _always_ apply regardless of the type. Do we end up with a weird split here where some impls are public API and others are not?

Another variant of this case:
- public API function / method / trait method returns a value with `doc(hidden)` type

#### `doc(hidden)` on non-defaulted associated items in unsealed public API trait (e.g. associated type / const / fn)

Perhaps surprisingly, public non-hidden unsealed traits can _sometimes_ safely have `doc(hidden)` associated items — if all those hidden associated items have defaults. That allows external implementers of the trait to do so without naming or changing the hidden associated items :+1:

But if the trait is public API, not sealed, and has associated items that are both hidden _and_ required, then we're in a catch-22: the trait *should* be possible to implement, but is not allowed to be implemented because doing so _requires_ touching non-public API items.

```rust
pub trait NonSealed {
    // Can't implement this trait without setting `NonSealed::T` here.
    #[doc(hidden)]
    type T;

    // Can't implement this trait without setting a value for this constant.
    #[doc(hidden)]
    const WITHOUT_DEFAULT: i64;

    // Can't implement this trait without implementing this function.
    #[doc(hidden)]
    fn implement_me();
}
```

In rare cases, macro crates might create traits that simultaneously:
- are intended to be implemented only in macro-generated code (so cannot be sealed) and nowhere else
- are intended to be _used_ as public API by user code, so cannot be `doc(hidden)` themselves

It's possible that such macro crates may consider this lint a false-positive. It's also possible that this could be worked around by defining two traits:
- One unsealed `doc(hidden)` trait for the macro-generated code to create impls for.
- One sealed public API trait that has a blanket impl for all implementers of the previous trait, which defines the public API surface area that users are allowed to interact with.

**Comment by kpreid** [comment]
Another use case for this lint would be for a package’s own tests of their public API to ensure that they themselves are not unintentionally making use of hidden items.
